File: leetcode-js/213-house-robber-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    'use strict';

    const dp1 = [0, nums[0]];
    const dp2 = [0, 0];

    for (let i = 1; i < nums.length; i += 1) {
        if (i === nums.length - 1) {
            dp1[i+1] = Math.max(dp1[i-1], dp1[i]);
        } else {
            dp1[i+1] = Math.max(nums[i] + dp1[i-1], dp1[i]);
        }
        dp2[i+1] = Math.max(nums[i] + dp2[i-1], dp2[i]);
    }

    return Math.max(dp1[nums.length], dp2[nums.length]);
};

File: leetcode-java/213-house-robber-ii.java
public class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        int n = nums.length;
        int[] dp1 = new int[n + 1];
        int[] dp2 = new int[n + 1];

        dp1[0] = 0;
        dp1[1] = nums[0];

        dp2[0] = 0;
        dp2[1] = 0;

        for (int i = 1; i < n; i++) {
            if (i == n - 1) {
                dp1[i + 1] = Math.max(dp1[i - 1], dp1[i]);
            } else {
                dp1[i + 1] = Math.max(dp1[i - 1] + nums[i], dp1[i]);
            }
            dp2[i + 1] = Math.max(dp2[i - 1] + nums[i], dp2[i]);
        }

        return Math.max(dp1[n], dp2[n]);
    }
}
